File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution: 
    def maxProfit(self, prices: List[int]) -> int:
        # [7,1,5,3,6,4,0,9]
        if len(prices) < 2: return 0

        buy,sell = 0,1

        max_profit = 0

        while sell < len(prices):
            if prices[sell] > prices[buy]:
                profit = prices[sell] - prices[buy]
                max_profit = max(max_profit,profit)
            else:
                buy = sell
            sell += 1

        return max_profit

    # SC -> O(1)
    # TC -> O(N)

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        if(prices.length == 0) return prices[0];

        int maxProfit = 0;
        int buy = 0;
        int sell = 1;

        while(sell < prices.length){
            if(prices[sell] <= prices[buy]){
                buy = sell;
            } else{
                int profit = prices[sell] - prices[buy];
                maxProfit = Math.max(maxProfit,profit);
            }
            sell+=1;
        }

        return maxProfit;
        
    }
}